crates/bitwarden/src/tool/generators/username.rs

```diff
@@ -0,0 +1,237 @@
+        }
+        Subaddress { r#type, email } => Ok(username_subaddress(&mut thread_rng(), r#type, email)),
+        Catchall { r#type, domain } => Ok(username_catchall(&mut thread_rng(), r#type, domain)),
+        Forwarded { service, website } => {
+            use crate::tool::generators::username_forwarders::*;
+            use ForwarderServiceType::*;
+            match service {
+                AddyIo {
+                    api_token,
+                    domain,
+                    base_url,
+                } => addyio::generate(http, api_token, domain, base_url, website).await,
+                DuckDuckGo { token } => duckduckgo::generate(http, token).await,
+                Firefox { api_token } => firefox::generate(http, api_token, website).await,
+                Fastmail { api_token } => fastmail::generate(http, api_token, website).await,
+                ForwardEmail { api_token, domain } => {
+                    forwardemail::generate(http, api_token, domain, website).await
+                }
+                SimpleLogin { api_key } => simplelogin::generate(http, api_key, website).await,
+            }
+        }
+    }
+}
+
+fn username_word(mut rng: impl RngCore, capitalize: bool, include_number: bool) -> String {
+    let word = EFF_LONG_WORD_LIST
+        .choose(&mut rng)
+        .expect("slice is not empty");
+
+    let mut word = if capitalize {
+        capitalize_first_letter(word)
+    } else {
+        word.to_string()
+    };
+
+    if include_number {
+        word.push_str(&random_number(&mut rng));
+    }
+
+    word
+}
+
+fn username_subaddress(mut rng: impl RngCore, r#type: AddressType, email: String) -> String {
+    if email.len() < 3 {
+        return email;
+    }
+
+    let (email_begin, email_end) = match email.find('@') {
+        Some(pos) if pos > 0 && pos < email.len() - 1 => {
+            email.split_once('@').expect("The email contains @")
+        }
+        _ => return email,
+    };
+
+    let email_middle = match r#type {
+        AddressType::Random => random_lowercase_string(&mut rng, 8),
+        AddressType::WebsiteName { website } => website,
+    };
+
+    format!("{}+{}@{}", email_begin, email_middle, email_end)
+}
+
+fn username_catchall(mut rng: impl RngCore, r#type: AddressType, domain: String) -> String {
+    if domain.is_empty() {
+        return domain;
+    }
+
+    let email_start = match r#type {
+        AddressType::Random => random_lowercase_string(&mut rng, 8),
+        AddressType::WebsiteName { website } => website,
+    };
+
+    format!("{}@{}", email_start, domain)
+}
+
+fn random_number(mut rng: impl RngCore) -> String {
+    let num = rng.gen_range(0..=9999);
+    format!("{num:0>4}")
+}
+
+fn random_lowercase_string(mut rng: impl RngCore, length: usize) -> String {
+    const CHARSET: &[u8] = b"abcdefghijklmnopqrstuvwxyz1234567890";
+    (0..length)
+        .map(|_| (*CHARSET.choose(&mut rng).expect("slice is not empty")) as char)
+        .collect()
+}
```

Comment:
Should we try and build up some standard distribution for use with `rng`?

Reply:
Good idea, done